rfd/0030-client-side-session-encryption.md

```diff
@@ -0,0 +1,155 @@
+### Configuration
+
+In order to activate client-side encryption, the user will specify 256-bit encryption key in the teleport config file. 
+
+The key will be set in the `teleport/storage` block. The value may be either a hex string, or a path to a file _containing_ a hex encoded key.
+
+```yaml
+teleport:
+  storage:
+    audit_sessions_uri: s3://teleport/recordings?region=ap-southeast-2
+    audit_sessions_local_key: "020a870107feb729b6795f6248a592e5332f71a2981455f91770d0c48096bca9"
+```
+    
+If the key is supplied, Teleport will use client-side encryption, with the supplied key acting as the master key. 
+
+If no key is supplied, Teleport will use the `s3manager` upload  & download clients. Any server-side encryption is largely dependant on the target bucket settings and is outside the scope of this document. 
+
+### Key Wrapping Algorithm
+
+The AWS SDK uses interfaces defined in the `s3crypto` package to define key wrapping algorithms. If we provide a compatible key wrapping and unwrapping mechanism, we _should_ be able to use the existing SDK machinery to create and retrieve client-side-encrypted object in a manner compatible with other clients.
+
+The Java SDK uses the well-defined [RFC3394 AESWrap](https://datatracker.ietf.org/doc/html/rfc3394) key wrapping algorithm. [Several](https://pkg.go.dev/github.com/chiefbrain/ios/crypto/aeswrap) [implementations](https://pkg.go.dev/github.com/dunhamsteve/ios/crypto/aeswrap) [of RFC3394](https://pkg.go.dev/github.com/gwatts/ios/crypto/aeswrap) in Go already exist.
+
+Integrating the key wrapping algorithm would require
+ 1. Selecting an implementation (or implementing our own, but probably not) of RFC3394
+ 2. Wrapping the RFC in a type 
+   * that knows the master key, and 
+   * implements the `s3crypto.ContentCipherBuilder` and `s3crypto.CipherDataDecrypter` interfaces
+ 3. Registering the `CipherDataDecrypter` in an [`s3crypto.CryptoRegistry`](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3crypto/#CryptoRegistry) so that the S3 SDK can use it to decrypt objects on download.
+
+### Integration
+
+The actual integration build down to a straightforward exercise in abstraction:
+
+1. Define a common interface for session uploading and downloading
+2. Wrap the Key wrapper, Crypto registry, the en- & decryption clients, etc into something that exposes that interface
+3. Wrap the existing `s3manager` in something that exposes the same interface
+4. Swap out the implementation based on the existence of the master key in the config file
+
+## Alternative Implementations
+
+### Custom KeyPadder
+
+If we are not interested in preserving compatibility with the AWS SDK for other languages, we can easily define a custom key wrapeer using the PCKS7 padder in the AWS SDK. This would remove the need for the AESWrap implementation, but would also make the encryption format Teleport-specific. Other AWS clients would not be able to interpret the at all.
```

Comment:
```suggestion
If we are not interested in preserving compatibility with the AWS SDK for other languages, we can easily define a custom key wrapper using the PCKS7 padder in the AWS SDK. This would remove the need for the AESWrap implementation, but would also make the encryption format Teleport-specific. Other AWS clients would not be able to interpret the encrypted session data at all.
```

Reply:
Obsoleted in rewrite